tucan/graph_utils.py

```diff
@@ -49,8 +48,22 @@ def permute_molecule(m, random_seed=1.0):
     return m_permu
 
 
+def _sorted_labels(m):
+    nodes = list(m.nodes(data=True))
+    n = nx.Graph()
+    n.add_nodes_from(sorted(nodes))
+    n.add_edges_from(m.edges(data=True))
+    return n
+
+
 def _permute_molecule(m):
     labels = list(m.nodes)
     permuted_labels = list(labels)  # shallow copy
     random.shuffle(permuted_labels)
-    return relabel_molecule(m, permuted_labels, labels)
+    relabeled_graph = relabel_molecule(m, permuted_labels, labels)
+
+    # NetworkX's relabel_nodes function does not change the iteration order of
```

Comment:
Remember: Before this PR `test_invariance` did not fail for semiheavy water, which was kind of odd.

With these changes `test_invariance` will fail for semiheavy water in case the isotope mass is not taken into account for the canonicalization.

Reply:
I like the `invariant code definition` in `def canonicalize_molecule()`:

    invariant_code_definitions = [
            _invariant_code_definition("atomic_number"),
            _invariant_code_definition("mass", 0),
            _invariant_code_definition("rad", 0),
        ]
        
Did I get this correctly that one could easily extend the canonicalization procedure to consider additional invariants by just adding extra lines in the data structure above and the order is also controlled by the listing, with invariant codes higher up given priority over those below?

Reply:
Exactly!

Reply:
Sweet - but still leave it to Jan to approve the pull request since I'm not deep enough into the details of this part of the code